### PR TITLE
8254807: Optimize startsWith() for String.substring()

### PR DESCRIPTION
**Pull request by navyxliu**
The optimization transforms code from s=substring(base, beg, end); s.startsWith(prefix)
to substring(base, beg, end) | base.startsWith(prefix, beg). 

it reduces uses of substring. hopefully c2 optimizer can remove the used substring.
/cc hotspot-compiler hotspot

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8254807](https://bugs.openjdk.java.net/browse/JDK-8254807): Optimize startsWith() for String.substring() ⚠️ Issue is not open.

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/974/head:pull/974`
`$ git checkout pull/974`

### CODE REVIEW

**Comment by navyxliu** [comment]
```
$make test TEST="micro:SubstringAndStartsWith"  CONF=linux-x86_64-server-release

Benchmark                                         (substrLength)   Mode  Cnt       Score      Error   Units
SubstringAndStartsWith.substr2StartsWith                       1  thrpt  100  140630.609 ?    9.404  ops/ms
SubstringAndStartsWith.substr2StartsWith                       8  thrpt  100  140634.988 ?    7.704  ops/ms
SubstringAndStartsWith.substr2StartsWith                      32  thrpt  100  135812.111 ? 4687.410  ops/ms
SubstringAndStartsWith.substr2StartsWith                     128  thrpt  100  140644.499 ?    7.856  ops/ms
SubstringAndStartsWith.substr2StartsWith                     256  thrpt  100  140640.309 ?    8.535  ops/ms
SubstringAndStartsWith.substr2StartsWith                     512  thrpt  100  140643.976 ?    8.168  ops/ms
SubstringAndStartsWith.substr2StartsWith_noalloc               1  thrpt  100  158672.117 ?    8.510  ops/ms
SubstringAndStartsWith.substr2StartsWith_noalloc               8  thrpt  100  158673.989 ?   10.198  ops/ms
SubstringAndStartsWith.substr2StartsWith_noalloc              32  thrpt  100  158677.660 ?    8.987  ops/ms
SubstringAndStartsWith.substr2StartsWith_noalloc             128  thrpt  100  153677.095 ? 5418.718  ops/ms
SubstringAndStartsWith.substr2StartsWith_noalloc             256  thrpt  100  158668.766 ?    9.138  ops/ms
SubstringAndStartsWith.substr2StartsWith_noalloc             512  thrpt  100  152544.827 ? 5922.157  ops/ms
Finished running test 'micro:SubstringAndStartsWith'
Test report is stored in build/linux-x86_64-server-release/test-results/micro_SubstringAndStartsWith
```

The micro demonstrates the optimization removes substring() and the throughput is constant value. 
it's still slight lower than handcrafted version `substr2StartsWith_noalloc` because the later doesn't need to check complete bound check.

File: test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 5, time = 200, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 20, time = 200, timeUnit = TimeUnit.MILLISECONDS)
+@State(Scope.Benchmark)
+public class SubstringAndStartsWith {
```
Comment (cl4es) [inline test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java:44]: I'd put this micro in org.openjdk.bench.java.lang and call it SubstringStartsWith

File: test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    @Param({"1", "8", "32", "128", "256", "512"})
```
Comment (cl4es) [inline test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java:45]: Does each param value pull its weight? I'd consider cutting down the default list to 2-3 variants (you can always specify more values on the command line et.c)

File: test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    @Benchmark
+    public boolean substr2StartsWith() {
+        return sample.substring(substrLength, substrLength * 2).startsWith(prefix);
+    }
+
+    @Benchmark
+    public boolean substr2StartsWith_noalloc() {
+        // compare prefix length with the length of substring
+        if (prefix.length() > substrLength) return false;
+        return sample.startsWith(prefix, substrLength); // substrLength here is actually the beginIdex of substring
```
Comment (cl4es) [inline test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java:68]:
```suggestion
        return sample.startsWith(prefix, substrLength); // substrLength here is actually the beginIndex of substring
```

**Review by cl4es — CHANGES_REQUESTED** [review]
Some comments and nits on the microbenchmark.

A general comment is that I think it would be good to add variants exercising UTF16 Strings: one where `sample` has some UTF-16 chars, and one where both `sample` and `prefix` do (latin-1 `sample` and UTF-16 `prefix` could be interesting too, to ensure this variant shortcuts quickly).

Should the `prefix` be something a bit more complex than a single char string? `startsWith("a", off)` is a case that'd be tempting to optimize down to `charAt(off) == 'a'` and then this micro might no longer do what it intends to do.

File: test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java (hunk 1 above)
Comment (navyxliu) [reply test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java:44]: got it.

File: test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java (hunk 2 above)
Comment (navyxliu) [reply test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java:45]: to be honest, I don't know the average length of substring in Java.   I add so many parameters to show this kind of pattern allocates and copies strings with  O(n) complexity, which n is length of substring.

to take your advice, I think I can model them in 3 representative lengths. how about this?
1.  small 4  like  a variable name
2.  medium size = 24 like a url or a filepath 
3. long string = 256   like a human-readable message

**Comment by navyxliu** [comment]
@cl4es 

Thank you for taking time to review this.  I understand you would like to see more variants, such as UTF16 strings and different prefixes.  

This api-level substitution actually doesn't care the underlying representation of string and prefix of startsWith. it works in the same way. The purpose of this microbench is to prove that substring() is not inevitable in a certain pattern.  JIT compilers can archive similar performance of the hand-craft code.  Right now, I have only a single variable, which is the length of substring. 
The result shows that the throughput is irrelevant of the lengths of substrings.  

My concern is that we would make results discernible if we introduce more than one variable. or I should write a group of benchmarks?

File: test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java (hunk 3 above)
Comment (navyxliu) [reply test/micro/org/openjdk/bench/vm/compiler/SubstringAndStartsWith.java:68]: thinks. I will fix it in next revision.

**Comment by navyxliu** [comment]
here is the result of updated micro benchmark. 
DoubeleBytes group is the utf-16 string.  SingleByte group the ascii string.
without the optimization,  the throughput is just 1/4~1/5 than hand-crafted code. 

With the optimization, the throughput can reach over 80%  of hand-crafted code.   The reason it can't reach 100% throughput of hand-crafted equivalence is that it has to generate stricter boundary checks.

```
// without OptimizeSubstring
Benchmark                                                  (substrLength)   Mode  Cnt       Score      Error   Units
SubstringStartsWith.substr2StartsWith_doubleBytes                       4  thrpt  125   40847.644 ? 1269.274  ops/ms
SubstringStartsWith.substr2StartsWith_doubleBytes                      24  thrpt  125   39284.961 ?  155.109  ops/ms
SubstringStartsWith.substr2StartsWith_doubleBytes                     256  thrpt  125   12370.385 ?  622.280  ops/ms
SubstringStartsWith.substr2StartsWith_singleByte                        4  thrpt  125   60633.544 ? 1989.512  ops/ms
SubstringStartsWith.substr2StartsWith_singleByte                       24  thrpt  125   60613.490 ? 1024.846  ops/ms
SubstringStartsWith.substr2StartsWith_singleByte                      256  thrpt  125   30212.033 ? 1106.752  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_doubleBytes               4  thrpt  125  154666.457 ?    7.132  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_doubleBytes              24  thrpt  125  154659.583 ?    7.663  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_doubleBytes             256  thrpt  125  154665.357 ?    6.414  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_singleByte                4  thrpt  125  162833.972 ?    8.170  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_singleByte               24  thrpt  125  162834.059 ?    7.862  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_singleByte              256  thrpt  125  162456.046 ?  217.304  ops/ms


// with OptimizeSubstring
Benchmark                                                  (substrLength)   Mode  Cnt       Score      Error   Units
SubstringStartsWith.substr2StartsWith_doubleBytes                       4  thrpt  125  119789.181 ? 3374.123  ops/ms
SubstringStartsWith.substr2StartsWith_doubleBytes                      24  thrpt  125  123740.637 ?   31.982  ops/ms
SubstringStartsWith.substr2StartsWith_doubleBytes                     256  thrpt  125  123701.525 ?   68.741  ops/ms
SubstringStartsWith.substr2StartsWith_singleByte                        4  thrpt  125  134529.257 ?    6.331  ops/ms
SubstringStartsWith.substr2StartsWith_singleByte                       24  thrpt  125  134517.222 ?    6.373  ops/ms
SubstringStartsWith.substr2StartsWith_singleByte                      256  thrpt  125  134527.929 ?    4.660  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_doubleBytes               4  thrpt  125  154668.784 ?    6.990  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_doubleBytes              24  thrpt  125  154567.971 ?   69.286  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_doubleBytes             256  thrpt  125  154630.115 ?   36.157  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_singleByte                4  thrpt  125  162852.894 ?    5.933  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_singleByte               24  thrpt  125  162850.630 ?    5.990  ops/ms
SubstringStartsWith.substr2StartsWith_noalloc_singleByte              256  thrpt  125  162841.513 ?    6.441  ops/ms
```

**Comment by navyxliu** [comment]
hello, 

May I ping to review this patch?  

Many business-oriented Java applications manipulate strings a lot.  Comparing to arithmetic scalar, string as oop is more expensive.  furthermore, many string objects need to be allocated on heap, so they increase workload of GC.  

By analyzing the construction of string object, I found that one source is String.substring().  My target is to eliminate substring  as many as possible.  This is the first attempt for me to enable substring optimization. If it works out, I will apply the api-substitution approach on other APIs such as String.charAt, StringBuilder::append, and even String.split(). 

thanks you in advance.

**Comment by navyxliu** [comment]
hi, Vladimir, 

Thank you for taking time to review it. 

I would like to call out that I plan to optimize substring().  My single aim to cut off a use of substring in this PR.
This is a gateway patch. I plan to apply a series of API substitutions for substrings. eg. StringBuilder::append, String.split and String.charAt etc. 

> _Mailing list message from [Vladimir Ivanov](mailto:vladimir.x.ivanov@oracle.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@openjdk.java.net):_
> 
> Hi Xin,
> 
> > the optimization transforms code from s=substring(base, beg, end); s.startsWith(prefix)
> > to substring(base, beg, end) | base.startsWith(prefix, beg).
> > it reduces an use of substring. hopefully c2 optimizer can remove the used substring()
> 
> It would be very helpful to see a more elaborate description of intended
> behavior to understand better what is the desired goal of the enhancement.
> 
> Though it looks attractive to address the problem in the JIT-compiler,
> it poses some new challenges which makes proposed approach questionable.
> I understand your desire to rely on existing String-related
> optimizations, but coalescing multiple concatenations differs
> significantly from your case.
> 
> Some of the concerns/questions I had while briefly looking through the
> patch:
> 
> - You introduce a call to a method (String::startsWith(String, int))
> which is not present in bytecode. It means (unless the method is called
> from different places) there won't be any profiling info present, and
> even if there is, it is unrelated to the call site being compiled. If
> the code is recompiled at some point (e.g., it hits an uncommon trap in
> startsWith(String,int) overload), there won't be any re-profiling
> happening. So, it can hit the very same condition on the consequent
> recompilations.
> 
Thank you to raise this! I didn't consider the profiling data before. 
it is an issue. Maybe it can explain why I observe only 80% performance of hand-craft code. 

TBH, I haven't had a clear and solid answer for it yet. 
The transformation actually won't alter control-flow. Perhaps, I can find a way to amend methoddata from the old method.
Or, is that possible I request to reprofile it after I modify some bytecodes?

> - "s.startsWith(prefix)" call node is reused and rewired to call
> unrelated method "base.startsWith(prefix, beg)". Will the new target
> method be taken into account during inlining? I would be much more
> comfortable seeing fresh call populated using standard process
> (involving CallGenerators et al) which then substitutes the original
> node. That way you make it less fragile longer term.
> 
I can make that happen. actually, I generated a brand-new call-generator in early revision of this patch. 
I drop it because it requires more code.

> - "hopefully c2 optimizer can remove the used substring()"
> If everything is inlined in the end, it can happen, but it's fragile.
> Instead, you could teach C2 that the method is "pure" (no interesting
> side effects to care about) and cut the call early. It already happens
> for boxing methods (see LateInlineCallGenerator::_is_pure_call for
> details).
> 
I do have the exact logic in early revision. I remarked substring() as late-inlining candidate and pure. 
I delete it because I found substring() is "always" inlined and c2 optimizer can do the deadcode elimination job if nobody uses it. 
I understand your concern.  I can improve it in the future. 

> Overall, if you want to keep the enhancement C2-specific, I'd suggest to
> look into intrinsifying String::startsWith(String, int) and not relying
> on its bytecodes at all. That way you would avoid fighting against the
> rest of the JVM in some situations.
> 
Because this optimization isn't one-off thing for startsWith, I can't intrinsify it. 
I'd like to explore an extensible approach.

> Best regards,
> Vladimir Ivanov

**Comment by navyxliu** [comment]
Vladimir is correct. I found one potential issue is that I can't inline "startsWith(String, int)" which is the overload of original "startsWith". _string_late_inlines still records the old callGenerator which associates with old callNode. I need to update the record and make sure the new callnode is considered in inline_string_calls(). working on it.

**Review by rose00 — COMMENTED** [review]
This is a lot of churn and a lot of code surface area to gain a small benefit.  I don't think this sort of change takes our code base in a healthy direction.  Is it really worth it?

Somebody deeply familiar with the existing string-builder optimizations needs to look at this carefully, and make sure that it is maintainable going forward into the future.  Once this integrates, we pay for the maintenance forever.

Even more basically, I think it may be an erroneous transformation.  Given `base = "abcd"` and `s = base.substring(base, 1, 2)` (i.e., `base = "b"`), the boolean `z1 = s.startsWith("bc")` is false even though the boolean `z2 = base.substring("bc", 1)` is true.  This means it is invalid to replace `z1` with `z2`.  If the optimization were to make this transform, deeply unpredictable things could happen in an application.

Now, I looked for the conditional that excludes this particular error case.  I didn't find it.  This tells me that one of two things is true:  (a) The proposed change is incorrect, or (b) the proposed change is so subtle it will be impossible to maintain in a state of correctness.

For these reasons, I am skeptical that (a) the change should be integrated, and (b) any similar change should be integrated.

Somebody who is deeply familiar with the existing string optimizations may have a different opinion, and I would be happy to accept their reassurances.

**Comment by vnkozlov** [comment]
I agree with John that these changes are moving in wrong direction.
I would like to know why Escape Analysis does not eliminate substring object if it is only used for StartsWith():
```
 public static boolean foo(String s) { 
     return s.substring(1).startsWith("a"); 
 }
```
Improving EA will get you more benefits.

**Comment by navyxliu** [comment]
Thanks for the feedbacks. They are extremely valuable for me. I am working on the issues that Vladimir Ivanov's comment about 'late inlining skips the overloaded methods" and the bug of "too short substring" from John's comment.

I just want to say that it's an opener PR. I try sell a technique I called it "API simplification".  From my observation, many java code can be more efficient if Java programmers just know "one more API".  Even they’ve known, we can't blame them for the hindsight because sometimes certain patterns such as s.substring(1).startsWith("a") only emerge after deep inlining.

Api simplification is here for help them out. I guess the real gain will be seen in this split case, even though it follows the same recipe here.
https://github.com/navyxliu/StringFunc/blob/master/src/com/amazon/jdkteam/brownbag/SplitAndPick.java#L57

For the startsWith case, I did spend some time to research it.  C2 can eliminate AllocateNode of newstring indeed, but I found c2 failed to eliminate two expensive nodes:  AllocateArray and ArrayCopy for the underlying byte buffer.  I don’t think this problem can be perfectly solved by EA and Scalar Replacement. Essentially, it’s redundancy and no new object should be allocated all all.

**Comment by navyxliu** [comment]
> _Mailing list message from [Vladimir Ivanov](mailto:vladimir.x.ivanov@oracle.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@openjdk.java.net):_
> 
> > > - You introduce a call to a method (String::startsWith(String, int))
> > > which is not present in bytecode. It means (unless the method is called
> > > from different places) there won't be any profiling info present, and
> > > even if there is, it is unrelated to the call site being compiled. If
> > > the code is recompiled at some point (e.g., it hits an uncommon trap in
> > > startsWith(String,int) overload), there won't be any re-profiling
> > > happening. So, it can hit the very same condition on the consequent
> > > recompilations.
> > 
> > 
> > Thank you to raise this! I didn't consider the profiling data before.
> > it is an issue. Maybe it can explain why I observe only 80% performance of hand-craft code.
> > TBH, I haven't had a clear and solid answer for it yet.
> > The transformation actually won't alter control-flow. Perhaps, I can find a way to amend methoddata from the old method.
> > Or, is that possible I request to reprofile it after I modify some bytecodes?
> 
> Sorry, I overlooked your question.
> 
> To be clear: the transformations during C2 compilation don't affect
> bytecodes in any way.
> 
> If a method is inlined, a deoptimization may lead to a surprising
> resumption of execution in the middle of unexpected method, but it's a
> one time occurrence and consequent executions will follow original path
> in the bytecode.
> 
Hi, Vladmir, 

I manage to inline the rewired method in c5a25bf.  I attempt to address your question "Will the new target
method be taken into account during inlining?".  From my benchmark, it is 10% faster than the last revision. 

Sorry, I still don't understand your concern about the inlined methods.  This is my first non-bugfix patch.  please forgive me if I look dumb.
If startsWith(String, int) is deoptimized, the control will give back to the interpreter. IMHO, I can resume execution back to the interpreter as long as jvms are all correct. I don't think I mess them up.  

> If inlining doesn't happen, then the JVM should be able to finish
> successully finish profiling in the callee if needed (reprofile request
> invalidates the nmethod and it forces the new call in C2 to go through
> either interpreter or C1 until it is compiled with C2).
> 
> So, I'm concerned specifically about inlined case. And I consider it a
> blocking issue for adapting the proposed approach of rewriting rules in
> C2. I don't see a way to reliably fix it without intrinsifying the
> operation.
>
I do mark substring and startsWith as intrinsics here. 
The reason I didn't mark "substring" pure because all nodes will be eliminated after late inlining if they are dead.
yes, if I mark it pure, it will drop the call node immediately in do_late_inline().
  
> Best regards,
> Vladimir Ivanov

In my defense, this patch is just 300 lines of change if you exclude test.  You can see that I try to reuse the existing codebase to minimize foosteps. As a beginner, I just deliberately make it reviewable-size. I don't think it's non-maintainable or a scramble patch. 

Yes, I admit that the gain is not so attractive so far. it's just open a possibility. A possibility of interprocedural optimization of java APIs.  I can also apply the same recipe from "StringBuilder.append(substring)" to "StringBuilder.append(base, begin, end)" etc. 

The bottom line is I have to learn in action.  All your feedbacks here are helping me grow. I really appreciate that and I think I can return it in other patches.  I plan to explore in Vladimir kolzov's direction next.  We can rewire the underlying bytebuffer of a string in substring pattern. Of course, it's only safe to do so if EA tells us that the object never escape.

**Comment by navyxliu** [comment]
hi, Vladimir, 

Now I understand what you mean.  vframeArray::unpack_to_stack() resumes vframe which points to the new overloaded method. It doesn't match the original bytecode and the entry of cpcache!

**Comment by vnkozlov** [comment]
@navyxliu during your investigation did you look on other applications in addition to `SpecJVM2008/xml.validation`?

I would like to know what patterns you see. For example, we would not benefit much if an extracted substring is used after startsWith() call. Also I don't think we should optimize Java code which can be replaced with existing String API as in RFE's example.

Note, in RFE's example you added length check which is not needed because startsWith(prefix, offset) does it:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/String.java#L1455

And if you see a lot of cases like `return substring(base, beg, end).startsWith(prefix)` or code where substring is used based on startsWith() check:
```
    String sub=substring(base, beg, end);
    if (sub.startsWith(prefix)) {
        // Do something with sub
```
we should consider extending String API with `startsWith(prefix, beg, end)`.

As always fixing Java code gives a lot more performance improvement than JIT compiler can.

**Comment by navyxliu** [comment]
> @navyxliu during your investigation did you look on other applications in addition to `SpecJVM2008/xml.validation`?
> 
> I would like to know what patterns you see. For example, we would not benefit much if an extracted substring is used after startsWith() call. Also I don't think we should optimize Java code which can be replaced with existing String API as in RFE's example.
> 
> Note, in RFE's example you added length check which is not needed because startsWith(prefix, offset) does it:
> https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/String.java#L1455
> 
> And if you see a lot of cases like `return substring(base, beg, end).startsWith(prefix)` or code where substring is used based on startsWith() check:
> 
> ```
>     String sub=substring(base, beg, end);
>     if (sub.startsWith(prefix)) {
>         // Do something with sub
> ```
> 
> we should consider extending String API with `startsWith(prefix, beg, end)`.
> 
> As always fixing Java code gives a lot more performance improvement than JIT compiler can.

Hi, Vladimir, 
Thank you for taking time to review it.  Besides xml.validation, I searched this pattern in our company-scope codebase.  I did find some cases, but it's not common indeed. Even xml.validation is unlikely to benefit from this PR.  To be honest, substring.startsWith is a special case in the middle of special cases.

My original aim is to eliminate object allocation and array copying of substrings if they are never escaped. I learned some unpopular APIs such as `String.startsWith(prefix, toOffset)`,  `StringBuilder.append(substr, toOffset)` and `String.split(sep, limit)` etc. I just feel I can add an API-level transformation to avoid creation of substrings in those case. It seems a cool trick. Given the fact that we need significant efforts to consolidate correctness, now I admit I am steering in wrong direction.  I closed JDK-8254807. 

I will go back to my original direction. I have extent current EA to support substring just like you did for boxing method. I think I can idealize() the Ideal pattern of String.substring().
